internal/datasource/git.go

```diff
@@ -386,7 +417,141 @@ func (s *GitSource) Changes(
 				Commit: foundRef.Hash().String(),
 			},
 		},
-	}, nil
+	}, ignore, nil
+}
+
+func (s *GitSource) changes(
+	ctx context.Context,
+	log hclog.Logger,
+	raw *pb.Job_DataSource,
+	baseDir string,
+	refNew, refCurrent string,
+	subpath string,
+) (bool, error) {
+	source := raw.Source.(*pb.Job_DataSource_Git)
+
+	// Normalize our subpath:
+	//   ./foo => foo
+	//   /foo  => foo
+	//
+	// Note that this SHOULD happen at the API level (we should reject
+	// any weird values) but we introduced validation later so we do this
+	// to clean up potentially old dirty values.
+	subpath = filepath.ToSlash(subpath)
+	if len(subpath) >= 2 && subpath[0] == '.' && subpath[1] == filepath.Separator {
+		subpath = subpath[2:]
+	} else if len(subpath) >= 1 && subpath[0] == filepath.Separator {
+		subpath = subpath[1:]
+	}
+
+	// Ensure our path ends with a '/' for comparison purposes later.
+	if !strings.HasSuffix(subpath, string(filepath.Separator)) {
+		subpath += string(filepath.Separator)
+	}
+
+	// Create a temporary directory where we will store the cloned data.
+	td, err := ioutil.TempDir(baseDir, "waypoint")
+	if err != nil {
+		return false, err
+	}
+	defer os.RemoveAll(td)
+
+	// Setup auth information
+	auth, err := s.auth(log, nil, source)
+	if err != nil {
+		return false, err
+	}
+
+	// Clone
+	var output bytes.Buffer
+	log.Trace("cloning repository", "url", source.Git.Url)
+	repo, err := git.PlainCloneContext(ctx, td, false, &git.CloneOptions{
+		URL:      source.Git.Url,
+		Auth:     auth,
+		Progress: &output,
+	})
```

Comment:
Yeah, it is very big. The way runners are currently written, we can’t really because each job gets its own directory that is blown away at the end of the job. I can try to pull in some caching mechanism here in this PR though.

The other option if this is opt-in is to just accept this, because this is happening ANYWAYS right now (we’re triggering a superfluous `up` which is forcing the checkout) and so at worst this is just making the current behavior better.

There is a “part 2” to the issue this fixes where 10 projects tracking the same repo will poll the repo 10 times, and it’d be nicer to actually poll 1 repo 1 time and trigger changes in 10 projects at once. That is fairly significant so not this PR, but I can also use that as a way to do caching cause we can just cache the “poll repo” operation.